File: app/src/main/java/com/akhbulatov/githubsample/ui/main/userdetailsroot/userdetails/UserDetailsFragment.java
package com.akhbulatov.githubsample.ui.main.userdetailsroot.userdetails;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.widget.Toolbar;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.akhbulatov.githubsample.App;
import com.akhbulatov.githubsample.R;
import com.akhbulatov.githubsample.models.UserDetails;
import com.akhbulatov.githubsample.ui.global.utils.CommonUtils;
import com.squareup.picasso.Picasso;

public final class UserDetailsFragment extends Fragment implements UserDetailsView {

    private static final String ARGUMENT_USER_LOGIN = "user_login";

    private ViewGroup contentLayout;
    private ImageView avatarImageView;
    private TextView loginValTextView;
    private TextView nameValTextView;
    private TextView locationValTextView;
    private TextView publicReposValTextView;
    private TextView followersValTextView;
    private TextView followingValTextView;
    private ViewGroup loadingProgressLayout;

    private UserDetailsPresenter presenter;
    private String login;
    @Nullable private UserDetails userDetails;

    public static UserDetailsFragment newInstance(String userLogin) {
        Bundle args = new Bundle();
        args.putString(ARGUMENT_USER_LOGIN, userLogin);

        UserDetailsFragment fragment = new UserDetailsFragment();
        fragment.setArguments(args);
        return fragment;
    }

    @Override public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Bundle args = getArguments();
        if (args == null) {
            throw new IllegalArgumentException("Must pass user login argument.");
        }
        login = args.getString(ARGUMENT_USER_LOGIN);
        presenter = new UserDetailsPresenter(App.getDataManager(), login);
    }

    @Nullable
    @Override
    public View onCreateView(
            @NonNull LayoutInflater inflater,
            @Nullable ViewGroup container,
            @Nullable Bundle savedInstanceState
    ) {
        return inflater.inflate(R.layout.fragment_user_details, container, false);
    }

    @Override public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        initViews(view);
        presenter.attachView(this);
    }

    @Override public void onDestroyView() {
        presenter.detachView();
        super.onDestroyView();
    }

    private void initViews(@NonNull View view) {
        setupToolbar(view);
        contentLayout = view.findViewById(R.id.contentLayout);
        avatarImageView = view.findViewById(R.id.avatarImageView);
        loginValTextView = view.findViewById(R.id.loginValTextView);
        nameValTextView = view.findViewById(R.id.nameValTextView);
        locationValTextView = view.findViewById(R.id.locationValTextView);
        publicReposValTextView = view.findViewById(R.id.publicReposValTextView);
        followersValTextView = view.findViewById(R.id.followersValTextView);
        followingValTextView = view.findViewById(R.id.followingValTextView);
        loadingProgressLayout = view.findViewById(R.id.loadingProgressLayout);
    }

    private void setupToolbar(@NonNull View view) {
        Toolbar toolbar = view.findViewById(R.id.toolbar);
        toolbar.setTitle(login);
        toolbar.setNavigationIcon(R.drawable.ic_arrow_back_white);
        toolbar.setNavigationOnClickListener(v -> presenter.onBackClicked());
        toolbar.inflateMenu(R.menu.user_details);
        toolbar.setOnMenuItemClickListener(menuItem -> {
                    switch (menuItem.getItemId()) {
                        case R.id.addToFavoritesMenu: {
                            presenter.onAddToFavoritesClicked(userDetails);
                            return true;
                        }
                        default:
                            return false;
                    }
                }
        );
    }

    @Override public void showContentLayout(boolean show) {
        contentLayout.setVisibility(show ? View.VISIBLE : View.GONE);
    }

    @Override public void showUserDetails(@NonNull UserDetails userDetails) {
        this.userDetails = userDetails;

        Picasso.get()
                .load(userDetails.avatarUrl)
                .placeholder(R.drawable.img_no_image)
                .into(avatarImageView);
        loginValTextView.setText(userDetails.login);
        nameValTextView.setText(userDetails.name);
        locationValTextView.setText(userDetails.location);
        publicReposValTextView.setText(String.valueOf(userDetails.publicRepos));
        followersValTextView.setText(String.valueOf(userDetails.followers));
        followingValTextView.setText(String.valueOf(userDetails.following));
    }

    @Override public void showLoadingProgress(boolean show) {
        loadingProgressLayout.setVisibility(show ? View.VISIBLE : View.GONE);
    }

    @Override public void showToFavoritesAdded() {
        CommonUtils.showToast(getContext(), R.string.user_details_added_to_favorites);
    }

    @Override public void backToUser() {
        if (getActivity() != null) {
            getActivity().finish();
        }
    }
}

File: app/src/main/java/com/akhbulatov/githubsample/ui/main/userdetailsroot/userdetails/UserDetailsFragment.kt
package com.akhbulatov.githubsample.ui.main.userdetailsroot.userdetails

import android.os.Bundle
import android.support.v4.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.akhbulatov.githubsample.App
import com.akhbulatov.githubsample.R
import com.akhbulatov.githubsample.models.UserDetails
import com.akhbulatov.githubsample.ui.global.utils.showToast
import com.squareup.picasso.Picasso
import kotlinx.android.synthetic.main.fragment_user_details.*
import kotlinx.android.synthetic.main.loading_error.*
import kotlinx.android.synthetic.main.loading_progress.*
import kotlinx.android.synthetic.main.toolbar.*

class UserDetailsFragment : Fragment(), UserDetailsView {

    private lateinit var presenter: UserDetailsPresenter
    private lateinit var login: String
    private var userDetails: UserDetails? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val args = arguments ?: throw IllegalArgumentException("Must pass user login argument.")
        login = args.getString(ARGUMENT_USER_LOGIN)
        presenter = UserDetailsPresenter(App.dataManager, login, App.errorHandler)
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? = inflater.inflate(R.layout.fragment_user_details, container, false)

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        setupToolbar()
        presenter.attachView(this)
    }

    override fun onDestroyView() {
        presenter.detachView()
        super.onDestroyView()
    }

    private fun setupToolbar() {
        toolbar.run {
            title = login
            setNavigationIcon(R.drawable.ic_arrow_back_white)
            setNavigationOnClickListener { presenter.onBackClicked() }
            inflateMenu(R.menu.user_details)
            setOnMenuItemClickListener { menuItem ->
                when (menuItem.itemId) {
                    R.id.addToFavoritesMenu -> {
                        userDetails?.let { presenter.onAddToFavoritesClicked(it) }
                        true
                    }

                    else -> false
                }
            }
        }
    }

    override fun showContentLayout(show: Boolean) {
        contentLayout.visibility = if (show) View.VISIBLE else View.GONE
    }

    override fun showUserDetails(userDetails: UserDetails) {
        this.userDetails = userDetails

        Picasso.get()
            .load(userDetails.avatarUrl)
            .placeholder(R.drawable.img_no_image)
            .into(avatarImageView)
        loginValTextView.text = userDetails.login
        nameValTextView.text = userDetails.name
        locationValTextView.text = userDetails.location
        publicReposValTextView.text = userDetails.publicRepos.toString()
        followersValTextView.text = userDetails.followers.toString()
        followingValTextView.text = userDetails.following.toString()
    }

    override fun showLoadingProgress(show: Boolean) {
        loadingProgressLayout.visibility = if (show) View.VISIBLE else View.GONE
    }

    override fun showError(error: String) {
        loadingErrorTextView.text = error
        loadingErrorLayout.visibility = View.VISIBLE
    }

    override fun showToFavoritesAdded() {
        showToast(R.string.user_details_added_to_favorites)
    }

    override fun backToUser() {
        activity?.finish()
    }

    companion object {
        private const val ARGUMENT_USER_LOGIN = "user_login"

        fun newInstance(userLogin: String): UserDetailsFragment {
            val args = Bundle().apply { putString(ARGUMENT_USER_LOGIN, userLogin) }
            return UserDetailsFragment().apply { arguments = args }
        }
    }
}
